plugins/woocommerce/src/Blocks/QueryFilters.php

```diff
@@ -120,6 +117,112 @@ public function get_stock_status_counts( $query_vars ) {
 		return $stock_status_counts;
 	}
 
+	/**
+	 * Get the count of on sale products.
+	 *
+	 * @param array $query_vars The WP_Query arguments.
+	 */
+	public function get_onsale_status_counts( array $query_vars ) {
+		$transient_key = 'wc_onsale_status_count_' . md5( wp_json_encode( $query_vars ) );
+		$cached_data   = get_transient( $transient_key );
+
+		if ( isset( $cached_data ) && ( ! defined( 'WP_DEBUG' ) || true !== WP_DEBUG ) ) {
+			return $cached_data;
+		}
+
+		global $wpdb;
+
+		add_filter( 'posts_clauses', array( $this, 'add_query_clauses' ), 10, 2 );
+		add_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$query_vars['no_found_rows']  = true;
+		$query_vars['posts_per_page'] = -1;
+		$query_vars['fields']         = 'ids';
+		$query                        = new \WP_Query();
+
+		$result            = $query->query( $query_vars );
+		$product_query_sql = $query->request;
+
+		remove_filter( 'posts_clauses', array( $this, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$count_sql = "
+			SELECT COUNT( DISTINCT product_id ) as status_count
+			FROM {$wpdb->wc_product_meta_lookup}
+			WHERE product_id IN ( {$product_query_sql} )
+			AND onsale > 0
+		"; // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+
+		$result = $wpdb->get_var( $count_sql ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+
+		$onsale_status_counts = array(
+			'onsale' => $result,
+		);
+
+		set_transient( $transient_key, $onsale_status_counts );
+
+		return $onsale_status_counts;
+	}
+
+	/**
+	 * Get the count of featured products.
+	 *
+	 * @param array $query_vars The WP_Query arguments.
+	 */
+	public function get_featured_status_counts( array $query_vars ) {
+		global $wpdb;
+
+		add_filter( 'posts_clauses', array( $this, 'add_query_clauses' ), 10, 2 );
+		add_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$query_vars['no_found_rows']  = true;
+		$query_vars['posts_per_page'] = -1;
+		$query_vars['fields']         = 'ids';
+		// phpcs:disable WordPress.DB.SlowDBQuery.slow_db_query_tax_query
+		$query_vars['tax_query'] = array(
+			'relation' => 'AND',
+			array(
+				'taxonomy' => 'product_visibility',
+				'field'    => 'name',
+				'terms'    => 'featured',
+			),
+			array(
+				'taxonomy' => 'product_visibility',
+				'field'    => 'name',
+				'terms'    => 'exclude-from-search',
+				'operator' => 'NOT IN',
+			),
+			array(
+				'taxonomy' => 'product_visibility',
+				'field'    => 'name',
+				'terms'    => 'exclude-from-catalog',
+				'operator' => 'NOT IN',
+			),
+		);
```

Comment:
There are some issues with this:
- this overrides the existing `tax_query` if exists.
- `exclude-from-search` and `exclude-from-catalog` queries should be handled by the product query that we're filtering from (main query or PC query). We should only add the featured tax query to the existing query.

Reply:
For example, here is the taxonomy query of the default catalog

<img width="290" alt="image" src="https://github.com/user-attachments/assets/2141e32d-011e-4b37-893e-7f3e44c81ca5">

Reply:
Resolving this convo since we're not implementing "Featured" in this PR.